compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+                    // To avoid introducing UB, just remove the StorageDead statements.
+                    storage_to_remove.insert(p.local);
+                }
+            }
+            Value::Unknown => {
+                // We do not have a value, so it's definitely not replacable.
+                fully_replacable_locals.remove(local);
+            }
+        }
+    }
+
+    Replacer { tcx, targets, fully_replacable_locals, storage_to_remove }
+        .visit_body_preserves_cfg(body);
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum LocationExtended {
+    Plain(Location),
+    Arg,
+}
+
+#[derive(Debug)]
+struct SsaLocals {
+    /// Assignments to each local.  This defines whether the local is SSA.
+    assignments: IndexVec<Local, Set1<LocationExtended>>,
+    /// Number of "direct" uses of each local, ie. uses that are not dereferences.
+    direct_uses: IndexVec<Local, u32>,
+    /// We visit the body in reverse postorder, to ensure each local is assigned before it is used.
+    /// We remember the order in which we saw the assignments to compute the SSA values in a single
+    /// pass.
+    assignment_order: Vec<Local>,
+}
+
+impl SsaLocals {
+    fn new(body: &Body<'_>) -> SsaLocals {
+        let direct_uses = IndexVec::from_elem(0, &body.local_decls);
+        let assignment_order = Vec::new();
+
+        let assignments = IndexVec::from_elem(Set1::Empty, &body.local_decls);
+        let mut this = SsaLocals { assignments, direct_uses, assignment_order };
+
+        for (local, decl) in body.local_decls.iter_enumerated() {
+            if matches!(body.local_kind(local), LocalKind::Arg) {
+                this.assignments[local] = Set1::One(LocationExtended::Arg);
+            }
+            // Do not bother tracking non-pointers.
+            if !decl.ty.is_any_ptr() {
+                this.assignments[local] = Set1::Many;
+            }
+        }
+
+        for (bb, data) in traversal::reverse_postorder(body) {
+            this.visit_basic_block_data(bb, data);
+        }
+
+        for var_debug_info in &body.var_debug_info {
+            this.visit_var_debug_info(var_debug_info);
+        }
+
+        debug!(?this.assignments);
+        debug!(?this.direct_uses);
+
+        this.assignment_order.retain(|&local| matches!(this.assignments[local], Set1::One(_)));
+        debug!(?this.assignment_order);
+
+        this
+    }
+
+    /// Compute the equivalence classes for locals, based on copy statements.
+    ///
+    /// The returned vector maps each local to the one it copies. In the following case:
+    ///   _a = &mut _0
+    ///   _b = _a
+    ///   _c = _a
+    ///   _d = _c
+    /// We return the mapping
+    ///   _a => _a // not a copy so, represented by itself
+    ///   _b => _a
+    ///   _c => _a
+    ///   _d => _a // transitively through _c
+    fn compute_copy_classes(&mut self, body: &Body<'_>) -> IndexVec<Local, Local> {
```

Comment:
why is this a method, but `fully_replacable_locals` and `compute_local_values` are not?